homeassistant/components/mqtt/client.py

```diff
@@ -657,36 +656,29 @@ async def _async_perform_subscriptions(self) -> None:
         # Since we do not know if a published value is retained we need to
         # (re)subscribe, to ensure retained messages are replayed
 
-        def _process_client_subscriptions() -> list[tuple[int, int]]:
-            """Initiate all subscriptions on the MQTT client and return the results."""
-            subscribe_result_list = []
-            for topic, qos in subscriptions.items():
-                result, mid = self._mqttc.subscribe(topic, qos)
-                subscribe_result_list.append((result, mid))
-                _LOGGER.debug("Subscribing to %s, mid: %s, qos: %s", topic, mid, qos)
-            return subscribe_result_list
-
-        subscriptions = self._pending_subscriptions
+        if not self._pending_subscriptions:
+            return
+
+        subscriptions: dict[str, int] = self._pending_subscriptions
         self._pending_subscriptions = {}
 
+        def _process_client_subscriptions() -> tuple[int, int]:
+            """Initiate all subscriptions on the MQTT client and return the results."""
+            result, mid = self._mqttc.subscribe(list(subscriptions.items()))
+            return result, mid
+
         async with self._paho_lock:
-            results = await self.hass.async_add_executor_job(
+            result, mid = await self.hass.async_add_executor_job(
                 _process_client_subscriptions
             )
```

Comment:
```suggestion
            subscription_list = list(subscriptions.items())
            result, mid = await self.hass.async_add_executor_job(
                self._mqttc.subscribe, subscription_list
            )
```

Reply:
I think we can drop the `_process_client_subscriptions` function since the return values are the same from subscribe now